File: ruby/simple-cipher/cipher.rb
class Cipher

  KEY_LENGTH = 100

  attr_reader :key
  def initialize(key = generate_key)
    raise ArgumentError if invalid_key?(key)
    @key = key
  end

  def encode(text)
    text.chars.map.with_index do |char, index|
      (wrap(char.ord + shift_ord(index))).chr
    end.join
  end

  def decode(text)
    text.chars.map.with_index do |char, index|
      (wrap(char.ord - shift_ord(index))).chr
    end.join
  end

  private

  def generate_key
    (0..KEY_LENGTH).map { ('a'..'z').to_a[rand(26)] }.join
  end

  def shift_ord(index)
    key[index].ord - 97
  end

  def wrap(val)
    val -= 26 if val > 122
    val += 26 if val < 97
    val
  end

  def invalid_key?(key)
    key.empty? ||
    key =~ /\d/ ||
    key =~ /[A-Z]/
  end

end

File: python/simple-cipher/cipher.py
import random
import string


class Caesar:
    @classmethod
    def encode(cls, phrase):
        return ''.join([chr(cls.wrap(ord(c) + 3)) for c in cls.clean(phrase)])

    @classmethod
    def decode(cls, phrase):
        return ''.join([chr(cls.wrap(ord(c) - 3)) for c in cls.clean(phrase)])

    @classmethod
    def wrap(cls, val):
        if val > 122:
            val -= 26
        elif val < 97:
            val += 26
        return val

    @classmethod
    def clean(cls, phrase):
        return filter(str.isalpha, phrase.lower())


class Cipher:

    def __init__(self, key=None):
        self.random_key_length = 100
        self.key = key if key is not None else self.generate_random_key()
        if not self.valid_key():
            raise ValueError()

    def encode(self, phrase):
        return ''.join([chr(self.wrap(ord(c) + self.offset(i))) for i, c in
                        enumerate(self.clean(phrase))])

    def decode(self, phrase):
        return ''.join([chr(self.wrap(ord(c) - self.offset(i))) for i, c in
                        enumerate(self.clean(phrase))])

    def wrap(self, val):
        while val > 122:
            val -= 26
        while val < 97:
            val += 26
        return val

    def clean(self, phrase):
        return filter(str.isalpha, phrase.lower())

    def offset(self, index):
        return self.wrap(ord(self.key[index % len(self.key)]) - 97)

    def generate_random_key(self):
        return ''.join(random.SystemRandom().choice(string.ascii_lowercase)
                       for _ in range(self.random_key_length))

    def valid_key(self):
        return self.key.isalpha() and self.key.islower()
